File: python/day12-testing-debugging/solutions/calculator.py
"""Day 12 reference solution -- type-hinted version of calculator.py."""


def add(a: float, b: float) -> float:
    return a + b


def subtract(a: float, b: float) -> float:
    return a - b


def divide(a: float, b: float) -> float:
    if b == 0:
        raise ValueError("cannot divide by zero")
    return a / b


def average(numbers: list[float]) -> float:
    if not numbers:
        raise ValueError("cannot average an empty list")
    return sum(numbers) / len(numbers)


def is_palindrome(text: str) -> bool:
    normalized = text.lower().replace(" ", "")
    return normalized == normalized[::-1]

File: typescript-javascript/day12-typescript-and-testing/calculator.ts
// The module under test for Day 12. Read through it -- everything here is
// already implemented and type-annotated, so you can focus on writing tests
// (and, in the last two functions, seeing interfaces/generics used for real).

export function add(a: number, b: number): number {
    return a + b;
}

export function subtract(a: number, b: number): number {
    return a - b;
}

export function divide(a: number, b: number): number {
    if (b === 0) {
        throw new Error("cannot divide by zero");
    }
    return a / b;
}

export function average(numbers: number[]): number {
    if (numbers.length === 0) {
        throw new Error("cannot average an empty array");
    }
    return numbers.reduce((sum, n) => sum + n, 0) / numbers.length;
}

export interface Student {
    name: string;
    scores: number[];
}

export function studentAverage(student: Student): number {
    return average(student.scores);
}

export function firstElement<T>(array: T[]): T | undefined {
    return array[0];
}
